src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -36,9 +36,186 @@ public class ProjectDependencyGraph
+                dependencySets: newDependencySets);
+        }
+
+        internal ProjectDependencyGraph WithAdditionalProjectReferences(ProjectId projectId, IEnumerable<ProjectId> referencedProjectIds)
+        {
+            // Ensure we only go over the list once, since we'll be enumerating it many times. This also gives us an easy bail if we can no-op
+            referencedProjectIds = referencedProjectIds.ToList();
+
+            if (!referencedProjectIds.Any())
+            {
+                return this;
+            }
+
+            // Step #1: recompute our references map
+            var newReferencesMap = _referencesMap;
+
+            if (newReferencesMap.TryGetValue(projectId, out var existingReferences))
+            {
+                newReferencesMap = newReferencesMap.SetItem(projectId, existingReferences.Union(referencedProjectIds));
+            }
+            else
+            {
+                newReferencesMap = newReferencesMap.SetItem(projectId, referencedProjectIds.ToImmutableHashSet());
+            }
+
+            // Step #2: compute our reverse map, if we had one previously to build off of
+            var newReverseReferencesMap = _lazyReverseReferencesMap;
+
+            if (newReverseReferencesMap != null)
+            {
+                var builder = newReverseReferencesMap.ToBuilder();
+
+                foreach (var referencedProject in referencedProjectIds)
+                {
+                    if (builder.TryGetValue(referencedProject, out var reverseReferences))
+                    {
+                        builder[referencedProject] = reverseReferences.Add(projectId);
+                    }
+                    else
+                    {
+                        builder[referencedProject] = ImmutableHashSet.Create(projectId);
+                    }
+                }
+
+                newReverseReferencesMap = builder.ToImmutable();
+            }
+
+            // Step #3: update our forward transitive map, by combining the transitive references of the new projects into the transitive references of projectId.
+            // If we didn't have transitive references computed for us, then we'll skip this because we won't be able to incrementally update what we don't have
+            var newTransitiveReferencesMap = _transitiveReferencesMap;
+
+            if (newTransitiveReferencesMap.TryGetValue(projectId, out var existingTransitiveReferences))
+            {
+                var builder = existingTransitiveReferences.ToBuilder();
+
+                // We already have some transitive references. We'll incorporate in transitive references of the projects we all have.
+                // If any of our dependencies haven't already computed their transitive references, we'll just remove the entry on newTransitiveReferencesMap so
+                // as to compute it lazily later.
+                foreach (var referencedProjectId in referencedProjectIds)
```

Comment:
Any particular concerns I can address?

Reply:
here's one idea: 

the part that gets confusing here is htat we're in instance methods computing new data, which we'll then use to create a totally new instance.   There's nothing per-se wrong with this.  But there's always this nagging concern in my head of: OMG, is he going to accidently mutate state of hte object.

I realize that's a somewhat silly concern as all the fields are readonly and Immutable objects.  But at teh same time, it still makes it a bit hard to reason about.

So a little suggestion i might have would be to call into some sort of static helper that computes all the new fields, and then construct the new value out of that.  Or, alternatively, have N static helpers that do nothing but interrogate the old data and return the new data for that particular field.  The top level algorithm is just:

```c#
var newXField = ComputeIncrementalUpdateToX(this.xField, args...)
var newYField = ComputeIncrementalUpdateToY(this.yField, args...)
...
return new Graph(new...);
```

This would also have hte added benefit of having each part of hte algorithm independent and easier to review.  Right now it's super hard to check to ensrue that you're not accidently using the wrong local vs. field in many of these sibling loops.

What do you think?

Reply:
Oh, I like that. Indeed at one point I did screw up and re-use a local, so 100% good suggestion.

Reply:
@CyrusNajmabadi doing so not only made this code a lot clearer, but also forced me to retest it further and realize I totally did screw up the transitive algorithms.